File: kotlin/mth-core/src/main/kotlin/mth/core/client/TelnetClientDelegate.kt
package mth.core.client

interface TelnetClientDelegate {
    /** Send raw bytes to the server socket. */
    fun write(data: ByteArray)

    /** Local echo state changed due to server ECHO negotiation. */
    fun onLocalEchoChanged(enabled: Boolean)

    /** GMCP has been negotiated. Called once when server's WILL GMCP is accepted. */
    fun onGMCPNegotiated() {}

    /** GMCP data received from server. */
    fun onGMCPReceived(module: String, json: String)

    /** MSDP variable update received from server. */
    fun onMSDPVariable(name: String, value: String)

    /** Server sent EOR or GA prompt marker. */
    fun onPromptReceived()

    /** MSSP data received from server. */
    fun onMSSPReceived(data: Map<String, String>) {}

    /** Server sent BEL (0x07) character. */
    fun onBellReceived() {}

    /** Log a diagnostic message. */
    fun log(message: String) {}
}

File: Sources/SwiftMTHClient/TelnetClientDelegate.swift
public protocol TelnetClientDelegate: AnyObject {
    /// Write raw bytes to the server connection.
    func write(data: [UInt8])

    /// Called when the server changes echo mode.
    /// When `enabled` is true, the client should echo locally.
    /// When false, the server is echoing (password mode).
    func onLocalEchoChanged(enabled: Bool)

    /// Called when GMCP subnegotiation completes.
    func onGMCPNegotiated()

    /// Called when a GMCP message is received from the server.
    func onGMCPReceived(module: String, json: String)

    /// Called when an MSDP variable is received from the server.
    func onMSDPVariable(name: String, value: String)

    /// Called when a prompt marker (GA or EOR) is received.
    func onPromptReceived()

    /// Called when a BEL character is received.
    func onBellReceived()

    /// Log a diagnostic message.
    func log(message: String)
}

public extension TelnetClientDelegate {
    func onGMCPNegotiated() {}
    func onBellReceived() {}
    func log(message: String) {}
}
